rfcs/ocamlnat.md

```diff
@@ -0,0 +1,140 @@
+
+# Motivation
+
+This work would provide a simple way to compile and execute OCaml code
+at runtime, which would unlock a lot of new possibilities to develop
+great new tools.
+
+Coupled with the fact that we can already embed cmi files into an
+executable, this work would make it possible to distribute a
+self-contained binary that can evaluate OCaml code at runtime. This
+would make it simple and straightforward to use OCaml as an extension
+language.
+
+## Verified  examples in documentation comment
+
+We are particularly interested in this feature for the [mdx][mdx]
+tool. More precisely, we are currently working on a feature allowing
+verified toplevel snippets in mli files. For instance:
+
+```ocaml
+(** [chop_prefix s ~prefix] returns [s] without the leading [prefix]. *)
+
+    {[
+      # chop_prefix "abc" ~prefix:"a";
+      - : string option = Some "bc"
+      # chop_prefix "abc" ~prefix:"x";
+      - : string option = None
+    ]}
+*)
+val chop_prefix : string -> prefix:string -> string
+```
+
+In the above example, the `{[ ... ]}` would be kept up to date by
+`mdx` to ensure that the document stays in sync when the code
+changes. In fact, the user would initially only write the `#` lines
+and `mdx` would insert the results just as with expectation tests.
+
+# The change in detail
+
+This change would add JIT code generation for x86 architectures as
+described in [the paper][paper]. For other architectures, we would
+still rely on the portable method of calling the assembler and linker.
+The main additions to the compiler code base would be:
+
+- some code in the backend to do the assembly in process
+- a few more C functions to glue things together
+
+The paper mentions that it adds 2300 lines of OCaml+C code to the
+compiler code base.
+
+One detail to mention: IIUC the [JIT ocamlnat][ocamlnat] from [the
+paper][paper] goes directly from the linear form to binary
+assembly. Now that we have a symbolic representation of the assembly
+we could also go from the symbolic assembly in order to share more
+logic between normal compilation and JIT.
+
+We discussed with @alainfrisch and @nobj since LexiFi has been using
+an in-memory assembler in production for a while. They mentioned that
+they would be happy to open-source the code if they can, which means
+that we could be using code that has been running in production for a
+long time and is likely to be well tested and correct.
+
+LexiFi's binary emitter is about 1800 lines of code including comments
+and newlines. This looks a bit smaller than the JIT part of the JIT
+ocamlnat, so we would still be adding approximately the same amount of
+code if we went this way.
+
+# Drawback
+
+This is one more feature to maintain in the compiler and it comes with
+a non-negligible amount of code. However, and especially if we can
+reuse LexiFi in-memory assembler, most of the additions would come from
+well tested code. @alainfrisch and @nobj also mentioned that this code
```

Comment:
```suggestion
well tested code. @alainfrisch and @nojb also mentioned that this code
```

Reply:
Oops, and twice as well :see_no_evil: Thanks for the fixes!